### PR TITLE
Touch keyboard working

### PR DESCRIPTION
**Pull request by MartijnBraam**
This makes the keyboard actually function, the coordinates on the keyboard get converted to the correct chars.

This makes it usable on phones without hardware keyboard if your password validates on this regex: `[0-9a-z ]+`, this is because caps and special chars arent implemented yet.

This depends on PR #9, implements issue #6

### CODE REVIEW

File: osk.conf [hunk 1]
```diff
@@ -5,4 +5,5 @@ wallpaper = #0099FF
 
 keyboard-background = #333333;
 keyboard-map = us
-keyboard-font = Roboto
+# keyboard-font = /usr/share/fonts/TTF/DroidSans.ttf
```
Comment (craftyguy) [inline osk.conf]: Why have this and comment it out? Is this path specific to pmos (whereas the one on line 8 is for your host OS) ?

Comment (MartijnBraam) [inline osk.conf]: The font path is specific to my dev machine, but for packaging in pmos we should probably just provide the config file in the abuild tree.

File: keyboard.cpp [hunk 2]
```diff
@@ -59,12 +114,42 @@ SDL_Surface *makeKeyboard(int width, int height) {
...
-  drawRow(surface, 0, rowHeight, width / 10, rowHeight, row2,
-          width / 100);
-  drawRow(surface, width / 20, rowHeight * 2, width / 10, rowHeight,
-          row3, width / 100);
+  if (TTF_Init() == -1) {
+    printf("TTF_Init: %s\n", TTF_GetError());
+    exit(1);
+  }
+
+  TTF_Font *font = TTF_OpenFont(config->keyboardFont.c_str(), 24);
```
Comment (craftyguy) [inline keyboard.cpp]: For some reason this is failing on the N900 when I set: `
keyboard-font = /usr/share/fonts/ttf-dejavu/DejaVuSansMono.ttf ` 


Here's the message printed, seems like it's not actually reading the config file for the path and instead using the "Dejavu" value in keyboard.cpp:

`TTF_OpenFont: Couldn't open DejaVu
`

File: keyboard.cpp [hunk 3]
```diff
@@ -59,12 +114,42 @@ SDL_Surface *makeKeyboard(int width, int height) {
...
   char row4[] = "zxcvbnm";
 
-  drawRow(surface, 0, 0, width / 10, rowHeight,row1, width / 100);
-  drawRow(surface, 0, rowHeight, width / 10, rowHeight, row2,
-          width / 100);
-  drawRow(surface, width / 20, rowHeight * 2, width / 10, rowHeight,
-          row3, width / 100);
+  if (TTF_Init() == -1) {
+    printf("TTF_Init: %s\n", TTF_GetError());
+    exit(1);
```
Comment (craftyguy) [inline keyboard.cpp]: We should really be calling `SDL_Quit()` here (and in all cases where we error out after calling SDL_Init), else this happens: 

` (!!!)  *** WARNING [Application exited without deinitialization of DirectFB!] *** [core.c:1639 in dfb_core_deinit_check()]
`

I don't know what the implications of quitting directfb without a clean exit, but I'd rather not find out!

File: keyboard.cpp (hunk 2 above)
Comment (MartijnBraam) [inline keyboard.cpp]: Did you specify the config file with `-c configfile`? DeJaVu is the fallback

**Comment by craftyguy** [comment]
Ah, I did not.. I assumed it would load the config automatically if it was in pwd. Maybe it should?

**Comment by craftyguy** [comment]
The keyboard doesn't seem to work on the N900's touchscreen. it seems that taps are producing relative positions (instead of absolute when running with `-t`. Here's some output when tapping on the N900's touchscreen on valid keys:

```
xMouse: 104     yMouse: -54                                                               
xMouse: 94      yMouse: -90                                                               
xMouse: 23      yMouse: 4
xMouse: 271     yMouse: -47
xMouse: -76     yMouse: -29
```

I've also noticed that the hardware keyboard registers two keypresses for hitting a button on the keyboard once (e.g. hitting 'a' on the N900 keyboard registers 'aa')

**Comment by MartijnBraam** [comment]
Hmm I don't think SDL handles the touchscreen calibration for the n900. The capacitive touchscreens are calibrated in the factory so they work fine without a calibration.

**Comment by craftyguy** [comment]
Ah, so what if the on-screen keyboard is not drawn for the N900 and only hardware key input is allowed? Would that be a good compromise?

**Comment by craftyguy** [comment]
Alternatively, I can build directfb/sdl2 to use tslib, which supports calibration!

**Comment by ollieparanoid** [comment]
Imho, if we roll our own sdl2/directfb anyway, we could add `tslib` as well. Then we support the use case of a non-keyboard phone with a resistive touch-screen, too!

**Comment by craftyguy** [comment]
@ollieparanoid 

Yep! I'm building tslib + directfb as we speak.

**Comment by craftyguy** [comment]
Ok it seems that the N900's resistive touchscreen works when using tslib, you just have to set `TSLIB_TSDEVICE=/dev/input/event3` prior to running osk-sdl.
I've update the (add_onscreenkeyboard branch)[https://github.com/postmarketOS/pmbootstrap/tree/feature/add_onscreenkeyboard] of `pmboostrap` to include tslib & directdb w/ tslib support.

**Pull request merged**